engine/src/witness/sol/program_swaps_witnessing.rs

```diff
@@ -0,0 +1,590 @@
+										source_address: Some(ForeignChainAddress::Sol(
+											sender.into(),
+										)),
+										channel_metadata: CcmChannelMetadata {
+											message: ccm_parameters
+												.message
+												.to_vec()
+												.try_into()
+												.map_err(|_| {
+													anyhow!(
+														"Failed to deposit CCM: `message` too long."
+													)
+												})?,
+											gas_budget: ccm_parameters.gas_amount.into(),
+											ccm_additional_data,
+										},
+									}),
+									vault_swap_parameters,
+								)
+							},
+						};
+						Ok(SolanaVaultSwapDetails {
+							from: from_asset,
+							deposit_amount: amount,
+							destination_address: EncodedAddress::from_chain_bytes(
+								dst_chain.try_into().map_err(|e| {
+									anyhow!("Error while parsing destination chain for solana vault swap:{}.", e)
+								})?,
+								dst_address.to_vec(),
+							)
+							.map_err(|e| {
+								anyhow!("Failed to decode the destination address for solana vault swap:{}.", e)
+							})?,
+							to: dst_token.try_into().map_err(|e| {
+								anyhow!("Error while decoding destination token for solana vault swap: {}.", e)
+							})?,
+							deposit_metadata,
+							swap_account: vault_swap_account,
+							creation_slot,
+							broker_fee,
+							affiliate_fees: affiliate_fees
+								.into_iter()
+								.map(|entry| cf_primitives::Beneficiary { account: entry.affiliate.into(), bps: entry.fee.into() })
+								.collect_vec()
+								.try_into()
+								.map_err(|_| {
+									anyhow!("runtime supports at least as many affiliates as we allow in cf_parameters encoding")
+								})?,
+							refund_params,
+							dca_params,
+							boost_fee,
+						})
+					};
+					(
+						VaultSwapAccountAndSender {
+							vault_swap_account,
+							swap_sender: sender.into(),
+						},
+						vault_swap_details()
+							.inspect_err(|e| {
+								warn!("Unable to derive swap details for account `{vault_swap_account}`: {e}")
+							})
+							.ok(),
+					)
+				}
+			},
+		)
+		.collect()
+		.await;
+
+	Ok((new_swaps, closed_accounts))
+
+	// TODO: When submitting data we could technically submit the slot when the SwapEvent was
+	// queried for the new opened accounts. However, it's just easier to submit the slot when the
+	// SwapEndpointDataAccount was queried for both closed accounts and new opened accounts.
+}
+
+async fn get_changed_program_swap_accounts(
+	sol_rpc: &SolRetryRpcClient,
+	sc_opened_accounts: HashSet<SolAddress>,
+	sc_closure_initiated_accounts: BTreeSet<VaultSwapAccountAndSender>,
+	swap_endpoint_data_account_address: SolAddress,
+) -> Result<(Vec<SolAddress>, Vec<VaultSwapAccountAndSender>, u64), anyhow::Error> {
+	let (_historical_number_event_accounts, open_event_accounts, slot) =
+		get_swap_endpoint_data(sol_rpc, swap_endpoint_data_account_address)
+			.await
+			.expect("Failed to get the event accounts");
```

Comment:
The expect message should state why we expect this not to fail. If it might fail, then we should throw the error up - it does look like it might fail, given we handle some different encodings, and we've seen some RPCs act weirdly on this, so I would say we throw an error to avoid panicking the whole engine

Reply:
also found another expect where we should also return an error. corrected both.